File: 929.Unique-Email-Addresses.py
# https://leetcode.com/problems/unique-email-addresses/description/
#
# algorithms
# Easy (80.4%)
# Total Accepted:    20.1k
# Total Submissions: 25.3k
# beats 98.82% of python submissions


class Solution(object):
    def numUniqueEmails(self, emails):
        """
        :type emails: List[str]
        :rtype: int
        """
        s = set()

        for email in emails:
            local, domain = email.split('@')
            plus_idx = local.find('+')
            if plus_idx != -1:
                local = local[:plus_idx]
            local = local.replace('.', '')

            s.add((local, domain))

        return len(s)

File: 929.Unique-Email-Addresses.java
// https://leetcode.com/problems/unique-email-addresses/
//
// algorithms
// Easy (68.08%)
// Total Accepted: 161,790
// Total Submissions: 237,662


class Solution {
    public int numUniqueEmails(String[] emails) {
        HashSet<String> set = new HashSet<>();

        for (String m : emails) {
            int idx = m.indexOf("@");
            String local = m.substring(0, idx);
            String host = m.substring(idx + 1);

            idx = local.indexOf("+");
            if (idx != -1) {
                local = local.substring(0, idx);
            }

            String[] tmp = local.split("\\.");
            StringBuilder sb = new StringBuilder();
            for (String s : tmp) {
                sb.append(s);
            }
            sb.append("@").append(host);

            set.add(sb.toString());
        }

        return set.size();
    }
}